src/main/scala/com/fulcrumgenomics/bam/AnnotateMolecularBarcodes.scala

```diff
@@ -0,0 +1,148 @@
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.bam
+
+import com.fulcrumgenomics.cmdline.{ClpGroups, FgBioTool}
+import com.fulcrumgenomics.util.ReadSegment.{MolecularBarcode, ReadSegment, Template}
+import com.fulcrumgenomics.util.ReadStructure
+import dagr.commons.CommonsDef.PathToBam
+import dagr.commons.io.Io
+import dagr.commons.util.LazyLogging
+import dagr.sopt._
+import dagr.sopt.cmdline.ValidationException
+import htsjdk.samtools._
+import htsjdk.samtools.util._
+
+import scala.collection.JavaConversions._
+
+@clp(description =
+  """
+    |Annotates reads with their molecular barcodes.  Only template bases will be retained as specified by the read
+    |structure.  If only one molecular barcode tag is given, then the molecular barcodes will be concatenated and stored
+    |in that tag.  Otherwise the number of molecular barcodes should match the number of tags.  Mapping information will
+    |be unchanged, as such, this tool should not be used on reads that have been mapped since it would lead to an BAM
+    |with inconsistent records.
+  """,
+  group = ClpGroups.SamOrBam)
+class AnnotateMolecularBarcodes
+( @arg(flag = "i",   doc = "Input BAM file.")                                     val input: PathToBam,
+  @arg(flag = "o",   doc = "Output BAM file.")                                    val output: PathToBam,
+  @arg(flag = "rs1", doc = "The read structure for read 1.")                       val readStructure1: String,
+  @arg(flag = "rs2", doc = "The read structure for read 2.")                       val readStructure2: String,
+  @arg(flag = "b1",  doc = "SAM tags in which to store the molecular barcodes (one-per segment) for read.")
+                                                                                  val molecularBarcodeTags1: Seq[String] = Seq.empty,
+  @arg(flag = "b2",  doc = "SAM tags in which to store the molecular barcodes (one-per segment).")
+                                                                                  val molecularBarcodeTags2: Seq[String] = Seq.empty,
+  @arg(flag = "a",   doc = "Annotate the read names with the molecular barcodes.") val annotateReadNames: Boolean = true,
+  @arg(flag = "t",   doc = "Annotate both reads from a read pair with each others tags.")
+                                                                                  val tagBothEnds: Boolean = true
+) extends FgBioTool with LazyLogging {
+
+  Io.assertReadable(input)
+  Io.assertCanWriteFile(output)
+  molecularBarcodeTags1.foreach(tag => if (tag.length != 2) throw new ValidationException("SAM tags must be of length two: " + tag))
+  molecularBarcodeTags2.foreach(tag => if (tag.length != 2) throw new ValidationException("SAM tags must be of length two: " + tag))
+  if (readStructure1.isEmpty) throw new ValidationException("readStructure1 may not be empty")
+  if (readStructure2.isEmpty) throw new ValidationException("readStructure2 may not be empty")
```

Comment:
Currently only PE bams are supported.  I would like to defer SE support until it is needed.